### ISSUE TITLE
maxdepth / iterations / files

### ISSUE THREAD

**Issue by Kolano** [PUNCH-Cyber/stoq-plugins-public#36]
A maximum_size variable is documented, does this plug-in not account for potentially malicious archives with:

- very deep layers of nesting
- very large numbers of tiny duplicate files
- etc

**Comment by mlaferrera** [comment]
Hi @Kolano, 

>very deep layers of nesting

The plugin itself does not account for this, however the stoQ framework itself does. I assume you are referring to attacks such as a zip bomb in this case. When instantiating [`Stoq()`](https://stoq-framework.readthedocs.io/en/latest/dev/core.html#api), you can define `max_recursion`. This will ensure that extracted payloads are not processed beyond the defined recursion value. The current default is `3`. You may also define `max_recursion` in the `stoq.cfg` file, or using the `stoq` command line argument `--max-recursion`.

> very large numbers of tiny duplicate files

While the plugin will not prevent this from happening, the framework will not scan duplicate extracted payloads. `Stoq()` [tracks the `sha1` hash](https://github.com/PUNCH-Cyber/stoq/blob/master/stoq/core.py#L520-L524) of each extracted payload. If it is a duplicate, it will simply move on to the next extracted payload. Once done, it will cleanup any remaining files from disk.

> etc

If there are more specific risks, I'd be happy to talk more about them. If they are not already mitigated, I'd be really curious to get your take on how to ensure the plugin and framework could be improved to address them.

Thanks for the questions. Please let me know if I misunderstood or didn't fully answer. I'd be very interested in getting your feedback.

**Comment by Kolano** [comment]
Thanks, that resolves my early concerns with archive analysis.

A few other things I'm unclear about:

- Handling VT search retries on hashes marked as queued for scanning, or if over API limits at the time of initial scans. I'm guessing I'd need to use a Deep Dispatcher to purge the entry from hashes_seen so it would be re-scanned, but I'm not clear. 
- General management of the scanned hashes inventory. It looks like it's maintained in an in memory set, and I'm not seeing anything to store it across sessions, pre-populate it on fresh installs, or otherwise maintain it.

**Comment by mlaferrera** [comment]
For VT search retries, I could improve the current plugin to better handle 50X responses from the VTMIS API. I'll add that as a todo.

For VTMIS API limit exceeded, the plugin will produce an error (which can be found in the stoQ response.) As for retry on API limit, the best way forward may be to check for the error in the results and queue it for processing again. This could be done using a decorator plugin (i.e., check for the error then send a message back to your provider plugin queue (i.e., Kafka, redis, pubsub) that indicates it needs to be rescanned. Let me know if this isn't very clear, I'd be happy to work with you on creating a plugin that would handle this use case.

As for scanned hash inventory, the `hashes_seen` variable only tracks hashes per payload chain. As an example, A `payload` is passed to `Stoq()` and 2 additional payloads are extracted from it. `hashes_seen` will only track 3 payloads to ensure there are no duplicates. Once the payload chain is complete, `hashes_seen` will reset. For the functionality you are looking for, this would be an ideal candidate for a plugin. 

I can see it either being a `DispatcherPlugin` or `ConnectorPlugin`. For example, a `payload` is passes to `Stoq()`, at which point the `DispatcherPlugin` (let's call it `dedup`), checks your data store (i.e., mongodb) to see if the hash exists. If the `payload` has already been scanned, it will not be scanned by any workers. If not, it will continue on with scanning. 

Another example could be using a `ConnectorPlugin`. The `payload` will be scanned as normal, but once it reaches the `ConnectorPlugin` before the plugin inserts into your datastore it checks if it already exists. If it does, it will either not insert anything, or increment a counter. Otherwise it will insert a new entry.

Neither one of the above plugins exist right now, but it should be simple enough to create or update one based on your needs. Again, I'm happy to help you develop a plugin to meet this use case.

I hope this help!

**Comment by Kolano** [comment]
Thanks so much for the rapid responses here.

Summary:

- Work may be needed to handle files queued for scanning by VT more cleanly. (This might be addressable alternate, per the next bullet, since I guess the abbreviated "queued for processing" response from VT is still returned)
- I'd need to generate worker plug-ins to pull details for payloads from DBs, or metadata files. I'm presuming an initial set of worker plugins could handle these tasks, with actual analysis tasks then assigned via a Deep Dispatcher, though I don't think publicly shared plugins for such exist.

One other query, perhaps I'm missing something, but it is stoQ multi-threaded? The only place I see "multiprocessing" is in the .cfg file...
```
# What is the maximum size of the multiprocessing queue?
# max_queue: 100
```
...if the current dispatchers rapidly receive files are they queued and processed one at a time, or would they be processed separately at the same time? Are there controls for handling such if it's the later.

**Comment by mlaferrera** [comment]
I believe your summary is accurate. One of the nice things about stoQ is that it is completely flexible because of the plugin architecture. For example, another way to check for duplicate payloads would be to write/modify a `ProviderPlugin` that checks your datastore for the payload before it is passed to `Stoq()` for processing. If it's a duplicate, it would simply skip the payload, otherwise it is passed along for processing.

The term `multiprocessing` in `stoq.cfg` is a relic from stoQ v1, sorry about that. I've fixed this in master. The current version does not support multiprocessing, but it is planned for a future release. stoQ however is multi-threaded, which is what `max_queue` refers to. As such, each file would be processed FIFO and one at a time per thread.

**Comment by Kolano** [comment]
OK. I'm guessing if I'll want it rather than using the dirmon provider I could launch multiple instances of stoQ on separate files.

**Comment by mlaferrera** [comment]
This is the first time I've heard of someone using incron to start a scan. It's really interesting to hear about the different use cases. In that case, you could skip the `ProviderPlugin` altogether and create a wrapper for `Stoq()`. For example, you can have incron call the below example script to check for previously scanned payloads (please don't use, this is really just meant as an example):

```python
import sys
import hashlib

payload = open(sys.argv[1], 'rb').read()
payload_hash = hashlib.sha1(payload).hexdigest()
hashes_seen = open('hashes.txt').read().splitlines()

if payload_hash not in hashes_seen:
  s = Stoq(...)
  response = s.scan(content, PayloadMeta(extra_data={'filename': sys.argv[1]}))
```

**Comment by Kolano** [comment]
You forgot...
`from stoq import Stoq`

**Comment by Kolano** [comment]
Back on the subject of the Decompress plug-in.

Archives with content too large for the disk result in...
```
    "errors": {
        "decompress": [
            "Unable to access extracted content"
        ]
    },
```
...which is true I guess, but really it wasn't even able to extract the contents. It would be preferable to first read archive meta-data...
`7z l file`
...and then extract if possible and still provide meta-data details if not.

**Comment by mlaferrera** [comment]
If a compressed file is too large, the plugin should raise an [exception](https://github.com/PUNCH-Cyber/stoq-plugins-public/blob/master/decompress/decompress/decompress.py#L111-L113), not the error you are seeing. The error you are seeing indicates something went wrong with opening the extracted object. I've pushed up a change that will provide more context around what error was raised.

**Comment by Kolano** [comment]
I'll check with the revised version, but it's the error I see with the .vhd file in this zip with an "application/octet-stream" mime type.

[sample12.zip](https://github.com/PUNCH-Cyber/stoq-plugins-public/files/2790330/sample12.zip)

**Comment by Kolano** [comment]
Current versions of 7z also support the following MIME types...

'application/vnd.debian.binary-package': '7z',
'application/vnd.ms-cab-compressed': '7z',
'application/x-arj': '7z',
'application/x-lha': '7z'
'application/x-lzma': '7z'
'application/x-rpm': '7z'
'application/x-xz': '7z'

**Comment by Kolano** [comment]
And with the revised version I get...
```
    "errors": {
        "decompress": [
            "Unable to access extracted content: "
        ]
    },
```

**Comment by mlaferrera** [comment]
Can you create a new issue for the decompression issue you're running into? This thread is getting a bit long with multiple discussions going, which may make it overly confusing for other users that may be experiencing the same issue.

As for the other formats, I have a commit that will be pushed shortly that will include them. Thanks again!

**Comment by Kolano** [comment]
- https://github.com/PUNCH-Cyber/stoq-plugins-public/issues/42
- https://github.com/PUNCH-Cyber/stoq-plugins-public/issues/43

**Comment by mlaferrera** [comment]
Thanks again @Kolano! I'm going to close this issue and move to the others.
